### PR TITLE
General Information - Allowed to be Partially Complete

### PR DESCRIPTION
**Pull request by jperson1**
# General Information - Allowed to be Partially Complete

Issue: https://github.com/GSA-TTS/FAC/issues/1992

## Changes:
1. The general information form can now be submitted partially complete.
    * HTML reflects this difference in that nothing is required anymore, and many error messages were struck.
    * JS reflects this difference in that no fields are made required, and null value errors have been removed.
    * The django form no longer requires all the fields, and has some regex validators to limit 400 responses from the gen info page
    * The validator (jsonnet schema) still runs, but it allows most fields to be empty.

2. The submission checklist page runs a different validator on the general information section
    * The new schema is `GeneralInformationComplete.jsonnet`, and it's similar to the other one, but it allows no empty data and requires all necessary fields.
    * When visiting the submission checklist page, you'll be able to tell when the form hasn't been totally completed.

3. Many test changes:
    * Some now need to be looking for errors in the page, rather than 400 responses
    * Some need to handle dates a little differently
    
4. Some other things:
    * The sidenav (top left) is fixed. It now has three sections that match up with the three major form groups. That broke at some point as we were updating the form. 

## How to test:
1. Switch to this branch and start a clean run. This will clear old submissions, which may or may not now be broken. 
2. Start a new submission, get to the general information form.
3. Verify that:
    * You can exit at this point and see your submission in the homepage
    * Visiting the checklist for that submission displays the form as incomplete
    * Navigating to the general information form still autocomplete some sections (fiscal year dates, UEI)
4. Fill out some fields and submit the form. Verify that:
    * The form still shows as incomplete
    * Those fields saved properly
    * Those fields auto-fill when revisiting the form.
5. Fill out some fields very wrong on purpose. Try to submit, and verify that these fields give reasonable error messages:
    * EINs (non-numbers, too long/short values)
    * Ciities (numbers, special characters like underscores)
    * Dates (not MM/DD/YYYY)
    * Phone numbers
6. Finish the form, submit, and verify that it displays as complete.
7. Play around with submitting/not submitting certain fields to try and find a case where it shows as complete when it isn't, or the opposite. 
8. Step through to cross-al and verify that general information gives errors if it isn't done, and gives no errors if it is done.

## Screenshots:
Submission progress page, incomplete general information:
image

Submission progress page, completed general information (the same as before):
image

General information with errors:
image

## PR checklist: submitters

- [x]   Link to an issue if possible. If there’s no issue, describe what your branch does. Even if there is an issue, a brief description in the PR is still useful.
- [x]   List any special steps reviewers have to follow to test the PR. For example, adding a local environment variable, creating a local test file, etc.
- [ ]   For extra credit, submit a screen recording like [this one](https://github.com/GSA-TTS/FAC/pull/1821).
- [x]   Make sure you’ve merged `main` into your branch shortly before creating the PR. (You should also be merging `main` into your branch regularly during development.)
- [x]   Make sure that whatever feature you’re adding has tests that cover the feature. This includes test coverage to make sure that the previous workflow still works, if applicable.
- [x]   Do manual testing locally. Our tests are not good enough yet to allow us to skip this step. If that’s not applicable for some reason, check this box.
- [ ]   Verify that no Git surgery was necessary, or, if it was necessary at any point, repeat the testing after it’s finished.
- [ ]   Once a PR is merged, keep an eye on it until it’s deployed to dev, and do enough testing on dev to verify that it deployed successfully, the feature works as expected, and the happy path for the broad feature area (such as submission) still works.

## PR checklist: reviewers

- [x]   Pull the branch to your local environment and run `make docker-clean; make docker-first-run && docker compose up`; then run `docker compose exec web /bin/bash -c "python manage.py test"`
- [x]   Manually test out the changes locally, or check this box to verify that it wasn’t applicable in this case.
- [x]   Check that the PR has appropriate tests. Look out for changes in HTML/JS/JSON Schema logic that may need to be captured in Python tests even though the logic isn’t in Python.
- [x]   Verify that no Git surgery is necessary at any point (such as during a merge party), or, if it was, repeat the testing after it’s finished.
        
The larger the PR, the stricter we should be about these points.

### CODE REVIEW

**Comment by tadhg-ohiggins** [comment]
I might have missed this part, but what indicates to the user the missing fields in the case where General Information has no errors but isn't finished?

**Comment by jperson1** [comment]
> what indicates to the user the missing fields in the case where General Information has no errors but isn't finished?

That's a great point. Nothing gives the user any particular missing field. We just have an "All fields are required" message up top. The only way the user knows the form is incomplete is because the link in the checklist doesn't turn green, and because cross-val will let them know directly if they get to it without completing the form. 

A potential quick fix is to add a conditional red asterisk (or other indicator) to the form so they can more easily see which fields are missing. Or, wrap them up and present the missing fields on the checklist or in the form somewhere.

**Comment by jperson1** [comment]
@tadhg-ohiggins From what I can tell, the validate will pass an error on the first field it runs into a problem on, but not on all the missing fields. For example, leaving the auditee address and city blank would throw an error on the address but not the city. Even if it's not friendly, this will still be great to have in corss-val, so unexpected validation issues can get caught. 

I wonder if a helper in submission_progress_check could walk it and check for missing fields, then display in the checklist? That wouldn't be as complex or comprehensive as the schema errors.

**Comment by jperson1** [comment]
@tadhg-ohiggins I've merged your cross val change into this branch. Since the "GeneralInformation.schema.json" version in this branch is super permissive, I have it using the "GeneralInformationComplete.schema.json". In whatever way the required fields set is merged in, I assume we'll want the more strict schema running in cross validation. 

It works great! Like you said, it's not user friendly, but it's certainly better than giving no errors.
image

**Review by tadhg-ohiggins — APPROVED** [review]

File: backend/audit/cross_validation/submission_progress_check.py [hunk 1]
```diff
@@ -114,3 +120,26 @@ def get_num_findings(award):
...
+    """
+    Given a base "progress" dictionary and the general_info object from a submission,
+    run validations to determine its completeness. Then, return a dictionary with
+    "general_information" as the key and the progress as the value.
+    """
+    try:
+        is_general_info_complete = bool(
+            validate_general_information_complete_json(general_info)
+        )
+    except Exception as err:
```
Comment (jadudm) [inline backend/audit/cross_validation/submission_progress_check.py:135]:
```suggestion
    except ValidationError as err:
```

File: backend/audit/cross_validation/submission_progress_check.py [hunk 2]
```diff
@@ -114,3 +120,26 @@ def get_num_findings(award):
...
+        print("General information form not complete based on conditions:", err)
```
Comment (jadudm) [inline backend/audit/cross_validation/submission_progress_check.py:136]:
```suggestion
```

File: backend/report_submission/views.py [hunk 3]
```diff
@@ -202,12 +211,59 @@ def post(self, request, *args, **kwargs):
             raise PermissionDenied("You do not have access to this audit.") from err
         except ValidationError as err:
             message = f"ValidationError for report ID {report_id}: {err.message}"
-            print(message)
             raise BadRequest(message)
         except LateChangeError:
             return render(request, "audit/no-late-changes.html")
+        except Exception as err:
+            message = f"Unexpected error in GeneralInformationFormView post. Report ID {report_id}: {err}"
```
Comment (jadudm) [inline backend/report_submission/views.py:218]:
```suggestion
            message = f"Unexpected error in GeneralInformationFormView post. Report ID {report_id}"
```

**Review by jadudm — APPROVED** [review]
LGTM.

**Pull request merged**